src/test/functions/env.spec.ts

```diff
@@ -0,0 +1,352 @@
+      expect(env.parse("FOO='foo#bar'")).to.deep.equal({
+        envs: { FOO: "foo#bar" },
+        errors: [],
+      });
+      expect(env.parse('FOO="foo#bar"')).to.deep.equal({
+        envs: { FOO: "foo#bar" },
+        errors: [],
+      });
+    });
+
+    it("ignores leading/trailing spaces before the separator", () => {
+      expect(env.parse("FOO   =      foo")).to.deep.equal({
+        envs: { FOO: "foo" },
+        errors: [],
+      });
+      expect(env.parse('FOO   =  "foo"')).to.deep.equal({
+        envs: { FOO: "foo" },
+        errors: [],
+      });
+      expect(env.parse("FOO   =  'foo'")).to.deep.equal({
+        envs: { FOO: "foo" },
+        errors: [],
+      });
+    });
+
+    it("ignores comments", () => {
+      expect(
+        env.parse(`
+FOO=foo # comment
+# line comment 1
+# line comment 2
+BAR=bar # another comment
+`)
+      ).to.deep.equal({
+        envs: { FOO: "foo", BAR: "bar" },
+        errors: [],
+      });
+    });
+
+    it("ignores empty lines", () => {
+      expect(
+        env.parse(`
+FOO=foo
+
+
+BAR=bar
+
+`)
+      ).to.deep.equal({
+        envs: { FOO: "foo", BAR: "bar" },
+        errors: [],
+      });
+    });
+
+    it("catches invalid lines", () => {
+      expect(
+        env.parse(`
+BAR###
+FOO=foo
+// not a comment
+=missing key
+`)
+      ).to.deep.equal({
+        envs: { FOO: "foo" },
+        errors: ["BAR###", "// not a comment", "=missing key"],
+      });
+    });
+  });
+
+  describe("validateKey", () => {
+    it("accepts valid keys", () => {
+      const keys = ["FOO", "ABC_EFG", "A1_B2"];
+      keys.forEach((key) => {
+        expect(() => {
+          env.validateKey(key);
+        }).not.to.throw();
+      });
+    });
+
+    it("throws error given invalid keys", () => {
+      const keys = ["", "1F", "B=C"];
+      keys.forEach((key) => {
+        expect(() => {
+          env.validateKey(key);
+        }).to.throw("must start with");
+      });
+    });
+
+    it("throws error given reserved keys", () => {
+      const keys = [
+        "FIREBASE_CONFIG",
+        "FUNCTION_TARGET",
+        "FUNCTION_SIGNATURE_TYPE",
+        "K_SERVICE",
+        "K_REVISION",
+        "PORT",
+        "K_CONFIGURATION",
+      ];
+      keys.forEach((key) => {
+        expect(() => {
+          env.validateKey(key);
+        }).to.throw("reserved for internal use");
+      });
+    });
+
+    it("throws error given keys with a reserved prefix", () => {
+      expect(() => {
+        env.validateKey("X_GOOGLE_FOOBAR");
+      }).to.throw("starts with a reserved prefix");
+
+      expect(() => {
+        env.validateKey("FIREBASE_FOOBAR");
+      }).to.throw("starts with a reserved prefix");
+    });
+  });
+
+  describe("load", () => {
+    const createEnvFiles = (sourceDir: string, envs: Record<string, string>): void => {
+      for (const [filename, data] of Object.entries(envs)) {
+        fs.writeFileSync(path.join(sourceDir, filename), data);
```

Comment:
Idle thought, I wonder if we should bring in something like https://www.npmjs.com/package/mock-fs instead of actually writing these files for each test run.

Reply:
Works great!

Reply:
Reverted since `mock-fs` only supports Nodejs12 and above 😢